File: server2/feed.py
# !!!note to self: this model still has a drift denoted by mu which is NOT preferable!!!
# update: the drift can be removed by equating the mu to 0, making it a martigale. However, more research should be done before finalising this value

import time
import random
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt
import redis
import json


redis_feed = redis.Redis(host='localhost', port=6379, db=0)
# publish new music in the channel epic_music

def brownianMotion(T, mu, sigma, dt, S0):
    N = round(T/dt)
    t = np.linspace(0, T, N)
    W = np.random.standard_normal(size = N) 
    W = np.cumsum(W)*np.sqrt(dt) ### standard brownian motion ###
    X = (mu-0.5*sigma**2)*t + sigma*W 
    S = S_feed[-1]*np.exp(X) ### geometric brownian motion ###
    return S

S0 = 20000
S_feed = [S0]
while True:
    S = brownianMotion(2, 0, 0.1, 0.01, 20000)
    S_feed.append(round((S[0]),2))
    S_feed.pop(0)
    final_S = S_feed[-1]

    time.sleep(1)
    date_time = datetime.now()
    current_time = date_time.strftime("%m/%d/%Y, %H:%M:%S")
    feed = {
        'price': final_S,
        'timestamp': current_time
        }
    print(feed)
    redis_feed.publish('price feed', json.dumps(feed))
    stream_name="sintetik_feed"
    redis_feed.xadd(stream_name,feed)
    #plt.scatter(current_time, final_S, linestyle='--')
    # plt.pause(1)

plt.show()


# in terminal:
# cd server2
# python3 -m venv venv
# . venv/bin/activate


# in another terminal:
    # python test_feed_sender.py

File: server2/feed.js
require('dotenv').config();
const linspace = require('linspace');
const math = require('mathjs');
const Redis = require('ioredis');
const env =process.env;
const { queryByPromise } = require('./dbconfig/db');
const redis = new Redis({
  host:'redis',
  port:env.REDIS_PORT
});

function brownianMotion(T, mu, sigma, dt) {
  let N, t;
  S = [];
  W = [];
  X = [];
  N = Math.round(T / dt);
  t = linspace(0, T, N);
  // Draw random samples from a normal (Gaussian) distribution
  // generate random numbers where max = 1, min = -1 as mean = 0, std = 1:
  for(var i=0; i<N; i++) {
  W[i] = Math.random() * (1 - (-1)) + (-1);;
  }
  W = math.cumsum(W);
  for(var i=0; i<W.length; i++) {
    W[i] = W[i] * Math.sqrt(dt);
  }
  // standard brownian motion (wiener process)
  for(var i=0; i<t.length; i++) {
	t[i] = t[i] * (mu - 0.5 * Math.pow(sigma, 2));
  }

  for(var i=0; i<W.length; i++) {
    X[i] = t[i] + (sigma * W[i]);
  }

  for(var i=0; i<X.length; i++) {
	  S[i] = S_feed[S_feed.length - 1] * Math.exp(X[i]);
  }
  return S;
  // geometric brownian motion
}

let S0 = 20000;
let S_feed = [S0];


function send_feed() {

let S, feed, final_S;
S = brownianMotion(2, 0, 0.1, 0.01);
// append only the first element of the geometric brownian motion into the S_feed array:
S_feed.push(S[0].toFixed(2));
// delete the previous element (first element) in the array:
S_feed.shift();
// take the last element of the S_feed array
final_S = S_feed.slice(-1)[0];
let current_time = Math.floor(Date.now()/1000) ;
feed = {
  "price": final_S,
  "timestamp": current_time,
  "symbol_name": "Volatility 10 (1s)"
};

( async () => {
//publish price feed
await redis.publish("price feed", JSON.stringify(feed));

//store data to redis stream
await redis.xadd("price feed","MAXLEN","86400", "*","price",feed.price,"timestamp",feed.timestamp,"symbol_name",feed.symbol_name);

//store data to postgres
const my_query = {
  text:`CALL storeFeed($1,$2,$3);`,
  values:[feed.symbol_name,feed.price,feed.timestamp]
};

await queryByPromise(my_query);

})();

};
setInterval(send_feed, 1000);
